src/background_thread.c

```diff
@@ -0,0 +1,554 @@
+	assert(pthread_equal(pthread_self(),
+	    background_thread_info[thread_ind].thread));
+
+	return NULL;
+}
+
+/* Create a new background thread if needed. */
+bool
+background_thread_create(tsd_t *tsd, unsigned arena_ind) {
+	assert(have_background_thread);
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &background_thread_lock);
+
+	/* We create at most NCPUs threads. */
+	size_t thread_ind = arena_ind % ncpus;
+	background_thread_info_t *info = &background_thread_info[thread_ind];
+	assert(info == arena_background_thread_info_get(arena_get(tsd_tsdn(tsd),
+	    arena_ind, false)));
+
+	bool need_new_thread;
+	malloc_mutex_lock(tsd_tsdn(tsd), &info->mtx);
+	need_new_thread = background_thread_enabled() && !info->started;
+	malloc_mutex_unlock(tsd_tsdn(tsd), &info->mtx);
+	if (!need_new_thread) {
+		return false;
+	}
+
+	pre_reentrancy(tsd);
+	int err;
+	load_pthread_create_fptr();
+	if ((err = pthread_create(&info->thread, NULL,
+	    background_thread_entry, (void *)thread_ind)) != 0) {
+		malloc_printf("<jemalloc>: arena %u background thread creation "
+		    "failed (%d).\n", arena_ind, err);
+	}
+	post_reentrancy(tsd);
+
+	malloc_mutex_lock(tsd_tsdn(tsd), &info->mtx);
+	assert(info->started == false);
+	if (err == 0) {
+		info->started = true;
+		n_background_threads++;
+	}
+	malloc_mutex_unlock(tsd_tsdn(tsd), &info->mtx);
+
+	return (err != 0);
+}
+
+bool
+background_threads_enable(tsd_t *tsd) {
+	assert(n_background_threads == 0);
+	assert(background_thread_enabled());
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &background_thread_lock);
+
+	VARIABLE_ARRAY(bool, created, ncpus);
+	unsigned i, ncreated;
+	for (i = 0; i < ncpus; i++) {
+		created[i] = false;
+	}
+	ncreated = 0;
+
+	unsigned n = narenas_total_get();
+	for (i = 0; i < n; i++) {
+		if (created[i % ncpus] ||
+		    arena_get(tsd_tsdn(tsd), i, false) == NULL) {
+			continue;
+		}
+		if (background_thread_create(tsd, i)) {
+			return true;
+		}
+		created[i % ncpus] = true;
+		if (++ncreated == ncpus) {
+			break;
+		}
+	}
+
+	return false;
+}
+
+bool
+background_threads_disable(tsd_t *tsd) {
+	assert(!background_thread_enabled());
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &background_thread_lock);
+
+	pre_reentrancy(tsd);
+	for (unsigned i = 0; i < ncpus; i++) {
+		background_thread_info_t *info = &background_thread_info[i];
+		bool wait;
+
+		malloc_mutex_lock(tsd_tsdn(tsd), &info->mtx);
+		if (info->started) {
+			wait = true;
+			info->started = false;
+			pthread_cond_signal(&info->cond);
+		} else {
+			wait = false;
+		}
+		malloc_mutex_unlock(tsd_tsdn(tsd), &info->mtx);
+
+		if (!wait) {
+			continue;
+		}
+		void *ret;
+		if (pthread_join(info->thread, &ret)) {
+			return true;
+		}
+		assert(ret == NULL);
+		n_background_threads--;
+	}
+	post_reentrancy(tsd);
+	assert(n_background_threads == 0);
+
+	return false;
+}
+
+/* Check if we need to signal the background thread early. */
+void
+background_thread_interval_check(tsdn_t *tsdn, arena_t *arena,
```

Comment:
This function appears to operate as I'd expect, but IMO it should be called whenever the number of unused dirty pages accrued since the last epoch advance or signal reaches a new high water mark.